File: backend/app/modules/reality/evidence_builder.py
from datetime import datetime, timezone
from typing import Dict, Any, Optional
from app.schemas import PhysicalObservation, EvidenceItem


class EvidenceBuilder:
    """
    Constructs strongly-typed EvidenceItem and PhysicalObservation adhering to NIA shared domain contracts.
    Guarantees privacy: image stays on-device; only metadata and extracted observations are transferred.
    """

    @staticmethod
    def build_evidence(
        raw_text: str,
        confidence: float,
        evidence_id: Optional[str] = None,
        source: str = "camera/OCR",
        is_simulated: bool = False,
        is_blurry: bool = False,
        captured_at: Optional[datetime] = None
    ) -> EvidenceItem:
        eid = evidence_id or f"ev-ocr-{int(datetime.now(timezone.utc).timestamp())}"
        dt = captured_at or datetime.now(timezone.utc)
        return EvidenceItem(
            evidence_id=eid,
            source=source,
            snippet=raw_text or "No text detected",
            confidence=confidence,
            captured_at=dt,
            media_ref=None,  # Privacy invariant: raw images are not uploaded to backend
            metadata={
                "is_simulated": is_simulated,
                "is_blurry": is_blurry,
                "privacy_mode": "ON_DEVICE_ONLY"
            }
        )

    @staticmethod
    def build_physical_observation(
        raw_text: str,
        location: Optional[str],
        confidence: float,
        entity: str = "Final Presentation",
        obs_id: Optional[str] = None,
        source: str = "camera/OCR",
        observed_at: Optional[datetime] = None,
        is_simulated: bool = False
    ) -> PhysicalObservation:
        oid = obs_id or f"obs-phys-{int(datetime.now(timezone.utc).timestamp())}"
        dt = observed_at or datetime.now(timezone.utc)
        return PhysicalObservation(
            id=oid,
            source=source,
            entity=entity,
            location=location,
            raw_text=raw_text,
            confidence=confidence,
            media_ref=None,  # Privacy invariant: raw images stay on phone
            observed_at=dt,
            metadata={"is_simulated": is_simulated}
        )

File: frontend/src/adapters/ocr/evidenceBuilder.ts
import { PhysicalObservation, EvidenceItem } from '../../contracts';
import { ExtractionResult, OCRResult } from './types';

export class EvidenceBuilder {
  /**
   * Builds an EvidenceItem adhering strictly to the shared domain contract.
   * Privacy: Local image URI is kept as a local device reference only.
   */
  public static buildEvidenceItem(
    ocrResult: OCRResult,
    extraction: ExtractionResult,
    evidenceId?: string
  ): EvidenceItem {
    const id = evidenceId || `ev-ocr-${Date.now()}`;
    return {
      evidenceId: id,
      source: ocrResult.isSimulated ? 'camera/simulated_ocr' : 'camera/OCR',
      snippet: extraction.rawText || ocrResult.rawText || 'No text detected',
      confidence: extraction.confidence,
      capturedAt: ocrResult.capturedAt,
      // Local reference only - raw image stays on phone, never uploaded to backend
      mediaRef: ocrResult.metadata?.imageUri ? `device-local://${ocrResult.metadata.imageUri}` : undefined,
      metadata: {
        isSimulated: ocrResult.isSimulated,
        providerType: ocrResult.providerType,
        isAmbiguous: extraction.isAmbiguous,
        candidateRooms: extraction.candidateRooms,
        noticeType: extraction.noticeType,
        privacyMode: 'ON_DEVICE_ONLY',
      },
    };
  }

  /**
   * Builds a PhysicalObservation adhering strictly to the shared domain contract.
   */
  public static buildPhysicalObservation(
    ocrResult: OCRResult,
    extraction: ExtractionResult,
    targetEntity: string = 'Final Presentation',
    observationId?: string
  ): PhysicalObservation {
    const id = observationId || `obs-phys-${Date.now()}`;
    return {
      id,
      source: ocrResult.isSimulated ? 'camera/simulated_ocr' : 'camera/OCR',
      entity: targetEntity,
      location: extraction.location,
      rawText: extraction.rawText,
      confidence: extraction.confidence,
      mediaRef: ocrResult.metadata?.imageUri ? `device-local://${ocrResult.metadata.imageUri}` : undefined,
      observedAt: ocrResult.capturedAt,
      metadata: {
        isSimulated: ocrResult.isSimulated,
        providerType: ocrResult.providerType,
        isAmbiguous: extraction.isAmbiguous,
        candidateRooms: extraction.candidateRooms,
        previousLocation: extraction.previousLocation,
        noticeType: extraction.noticeType,
      },
    };
  }
}
